httpbp/error_logger_test.go

```diff
@@ -0,0 +1,92 @@
+package httpbp
+
+import (
+	"io"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/reddit/baseplate.go/prometheusbp/promtest"
+)
+
+func initBaseLogger(w io.Writer) *zap.Logger {
+	// Mostly copied from zap.NewExample, to make the log deterministic.
+	encoderCfg := zapcore.EncoderConfig{
+		MessageKey:     "msg",
+		LevelKey:       "level",
+		NameKey:        "logger",
+		EncodeLevel:    zapcore.LowercaseLevelEncoder,
+		EncodeTime:     zapcore.ISO8601TimeEncoder,
+		EncodeDuration: zapcore.StringDurationEncoder,
+	}
+	core := zapcore.NewCore(zapcore.NewJSONEncoder(encoderCfg), zapcore.AddSync(w), zap.DebugLevel)
+	return zap.New(core)
+}
+
+func TestHTTPServerLogger(t *testing.T) {
+	t.Run("no-panic", func(t *testing.T) {
+		defer promtest.NewPrometheusMetricTest(t, "25192", httpServerLoggingCounter, prometheus.Labels{
+			"upstream_issue": "25192",
+		}).CheckDelta(1)
+
+		logger, err := httpServerLogger(nil, initBaseLogger(io.Discard))
+		if err != nil {
+			t.Fatalf("httpServerLogger failed: %v", err)
+		}
+		// Make sure that empty map won't cause panics with matching log lines.
+		logger.Printf("Hello, golang.org/issue/25192!")
+	})
+
+	t.Run("suppressed", func(t *testing.T) {
+		defer promtest.NewPrometheusMetricTest(t, "25192", httpServerLoggingCounter, prometheus.Labels{
+			"upstream_issue": "25192",
+		}).CheckDelta(2)
+
+		var sb strings.Builder
+		logger, err := httpServerLogger(map[int64]float64{
+			25192: 0,
+		}, initBaseLogger(&sb))
+		if err != nil {
+			t.Fatalf("httpServerLogger failed: %v", err)
+		}
+		logger.Printf("Hello, golang.org/issue/25192!")
+		logger.Printf("Hello, golang.org/issue/25192!")
+		if str := sb.String(); strings.TrimSpace(str) != "" {
+			t.Errorf("Expected logs being suppressed, got %q", str)
+		}
+
+		sb.Reset()
+		logger.Printf("Hello, world!")
+		const want = `{"level":"warn","msg":"Hello, world!","from":"http-server"}`
+		if got := sb.String(); strings.TrimSpace(got) != want {
+			t.Errorf("Got %q, want %q", got, want)
+		}
+	})
+
+	t.Run("not-suppressed", func(t *testing.T) {
+		defer promtest.NewPrometheusMetricTest(t, "25192", httpServerLoggingCounter, prometheus.Labels{
+			"upstream_issue": "25192",
+		}).CheckDelta(1)
+
+		var sb strings.Builder
+		logger, err := httpServerLogger(map[int64]float64{
+			25192: 1,
+		}, initBaseLogger(&sb))
+		if err != nil {
+			t.Fatalf("httpServerLogger failed: %v", err)
+		}
+		logger.Printf("Hello, golang.org/issue/25192!")
+		if got, want := sb.String(), `{"level":"warn","msg":"Hello, golang.org/issue/25192!","from":"http-server"}`; strings.TrimSpace(got) != want {
+			t.Errorf("Got %q, want %q", got, want)
+		}
+
+		sb.Reset()
+		logger.Printf("Hello, world!")
+		if got, want := sb.String(), `{"level":"warn","msg":"Hello, world!","from":"http-server"}`; strings.TrimSpace(got) != want {
+			t.Errorf("Got %q, want %q", got, want)
+		}
```

Comment:
🔕 just a nit that you do this differently on line 63

Reply:
that's intentional. it's a long string so I actually prefer to have it in its own line as a `const`. but in this case I need it twice so a `const` no longer works.